### PR TITLE
Add @napi-rs/glob package

### PR DESCRIPTION
**Pull request by mohsen1**
This is a very rough draft for adding a new package that wraps Rust's glob crate. I'm aiming to teach myself Rust and NAPI stuff by doing this. You don't have to review it right now but if you have any helpful comments I would greatly appreciate it! 

Fixes: https://github.com/napi-rs/node-rs/issues/605

## TODO

* [x] Benchmark 
* [ ] Default export https://github.com/napi-rs/napi-rs/issues/1046

### CODE REVIEW

File: packages/glob/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      .unwrap_or(false),
+    require_literal_separator: options
+      .as_ref()
+      .map(|o| o.require_literal_separator.unwrap_or(false))
+      .unwrap_or(false),
+    require_literal_leading_dot: options
+      .as_ref()
+      .map(|o| o.require_literal_leading_dot.unwrap_or(false))
+      .unwrap_or(false),
+  };
```
Comment (mohsen1) [inline packages/glob/src/lib.rs:53]: All I want to do is to expose `MatchOptions` to the NodeJS side. Are there better alternatives?

File: packages/glob/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+  for entry in glob_with(&pattern, glob_options).expect("Failed to read Glob pattern") {
+    match entry {
+      Ok(path) => {
+        results.push(path.to_str().unwrap().to_string());
+      }
+
+      Err(e) => println!("{:?}", e),
```
Comment (mohsen1) [inline packages/glob/src/lib.rs:61]: Does napi-rs have a better way of handing errors?

File: packages/glob/benchmark/bench.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import b from 'benny'
+
+import { plus100 } from '../index'
+
+function add(a: number) {
+  return a + 100
+}
+
+async function run() {
```
Comment (mohsen1) [inline packages/glob/benchmark/bench.ts:9]: Planning to bench against npm's glob https://www.npmjs.com/package/glob

File: packages/glob/src/lib.rs (hunk 2 above)
Comment (Brooooooklyn) [reply packages/glob/src/lib.rs:61]:
```suggestion
      glob_with(&pattern, glob_options).map_err(|err| napi::Error::new(napi::Status::GenericFailure, format!("glob failed {}", err)))?.try_for_each(|entry| {
          let path = entry.map_err(|err| napi::Error::new(napi::Status::GenericFailure, format!("read entry failed {}", err)))?;
          results.push(path.to_str().map_err(|err| napi::Error::new(napi::Status::GenericFailure, format!("path can not be covered to string {}", err))?.to_string());
          Ok(())
       })?
```

File: packages/glob/src/lib.rs (hunk 1 above)
Comment (Brooooooklyn) [reply packages/glob/src/lib.rs:53]: No... If upstream crates doesn't support `serde` or `napi` you can only covert it by hand

**Comment by mohsen1** [comment]
@Brooooooklyn thank you for your responses

Question: is there a way of making this async? In the Rust land it is a sync operation but it can take some time so it would be ideal to unlock the Node.JS event loop while doing work.

**Comment by Brooooooklyn** [comment]
https://napi.rs/docs/concepts/async-task

File: packages/glob/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  pub(crate) pattern: String,
+  pub(crate) options: Option<GlobOptions>,
+}
+
+#[napi]
+impl Task for GlobTask {
+  // TODO: Can't figure out how to type `Output` and `JsValue` to be an array of strings
+  // For now the async task only returns one string!
+  type Output = String; // Vec<String>;
+  type JsValue = String; // JsObject;
```
Comment (mohsen1) [inline packages/glob/src/lib.rs:96]: @Brooooooklyn I couldn't figure out how to make this an array of string. There is no `JsArray` type that I could use.

**Comment by mohsen1** [comment]
@Brooooooklyn Added some bench and mine is slower! Any ideas where we can improve the performance on this?

```
  "glob":
    14 120 ops/s, ±1.43%   | fastest

  "@napi-rs/glob":
    12 553 ops/s, ±0.19%   | slowest, 11.1% slower
```

**Comment by mohsen1** [comment]
Fixed the async API and benchmark code. Of course napi would be faster! 

Screen Shot 2022-01-30 at 1 27 29 PM

**Pull request closed without merging**